Cache report (#136)
imapp-pl/gas-cost-estimator: docs/gas-cost-estimator.md

<img src="./gas_cost_estimator_doc_assets/evmone_perf_overhead.png" width="425"/> <img src="./gas_cost_estimator_doc_assets/geth_perf_overhead.png" width="425"/> 

#### Measure marginal

In the first step we use measure marginal approach to verify whether cache usage depend on opcode.

For evmone the total cache effectiveness is between `1e-7` and `3e-7`. 
Recall that these denotes misses ratios. 
With such low values, computations are executed almost entirely within caches and cache usage profiles are very similar for all opcodes.
For the full picture L1 ratio to LLC is between `0.0008` and `0.0016`. 
This means that L1 cache performance has huge impact on measurements.
For completeness, the branch prediction effectiveness varies between `0.0008` and `0.0012`. 
And it may be considered almost equal for all opcodes.

For geth the total cache effectiveness is between `0.00006` and `0.00014`. 
So in this case also, computations are executed almost entirely within caches and cache usage profiles are very similar for all opcodes.
For the full picture L1 ratio to LLC is between `0.012` and `0.018`. 

> This means again that L1 cache performance has huge impact on measurements but slighly lower than for evmone.

Reviewer: ```suggestion
This means again that L1 cache performance has a huge impact on measurements but is slightly lower than for evmone.
```

For completeness, the branch prediction effectiveness varies between `0.008` and `0.014`. 
And it may be considered almost equal for all opcodes. 

<img src="./gas_cost_estimator_doc_assets/evmone_perf_marginal_total_effectiveness.png" width="425"/> <img src="./gas_cost_estimator_doc_assets/geth_perf_marginal_total_effectiveness.png" width="425"/> 

#### Validation

> Next we verify how cache usage profiles change when random programs, closer to real world contracts, are executed.

Reviewer: ```suggestion
Next, we verify how cache usage profiles change when random programs, closer to real-world contracts, are executed.
```

For evmone the total cache effectiveness is between `3e-7` and `4.5e-6`.
For geth ththis is between `4e-6` and `1.5e-4`. 

> But note that high values are attained for 0 length programs. These are not 0 length in fact, but very short.

Reviewer: ```suggestion
But note that high values are attained for 0-length programs. These are not 0 lengths in fact, but very short.
```

For other programs the ratios significantly drop. 

> So computations are executed almost entirely in caches, also for these programs, regardless the the fact ratios rised compering to marginal programs. 

Reviewer: ```suggestion
So computations are executed almost entirely in caches, also for these programs, regardless of the fact ratios rised compering to marginal programs. 
```
